File: 0001_0100/0049/0049.go
package main

import (
	"fmt"
	"os"
)

func count(n int) int {
	s := 0
	for ; n > 0; n &= n - 1 {
		s++
	}
	return s
}

func main() {
	f, _ := os.Open("input.txt")
	q, _ := os.Create("output.txt")
	var s, t string
	fmt.Fscan(f, &s, &t)
	set := make(map[byte]int)
	for i := byte('0'); i <= '9'; i++ {
		set[i] = 1 << (i - '0')
	}
	for i := byte('a'); i <= 'g'; i++ {
		set[i] = 15 << (i - 'a')
	}
	set['?'] = 1023
	n := 1
	for i := 0; i < len(s) && n > 0; i++ {
		n *= count(set[s[i]] & set[t[i]])
	}
	fmt.Fprint(q, n)
}

File: 0001_0100/0049/0049.c
#include <stdio.h>

int count(int n)
{
	int s = 0;
	while (n)
		s++, n &= n - 1;
	return s;
}

int main()
{
	FILE *f = fopen("input.txt", "r"), *q = fopen("output.txt", "w");
	char s[10], t[10];
	int k = 1, p[104];

	fscanf(f, "%s%s", s, t);
	for (int i = '0'; i <= '9'; ++i)
		p[i] = 1 << i - '0';
	for (int i = 'a'; i <= 'g'; ++i)
		p[i] = 15 << i - 'a';
	p['?'] = 1023;
	for (int i = 0; s[i] && k; ++i)
		k *= count(p[s[i]] & p[t[i]]);
	fprintf(q, "%d", k);
	return 0;
}
